---
title: "Introduction to clonealign"
shorttitle: "Assigning single cell gene expression to clone-of-origin using clonealign"
author: 
- name: "Kieran R Campbell"
  affiliation: 
    - Department of Statistics, University of British Columbia
    - Department of Molecular Oncology, BC Cancer Agency
    - UBC Data Science Institute, University of British Columbia
  email: kicampbell@bccrc.ca
date: "October 2018"
package: clonealign
bibliography: bibliography.bib
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to clonealign}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = FALSE
)


suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(scater)
  library(dplyr)
})
```


# Overview

`clonealign` assigns cells measured using single-cell RNA-seq to their clones of origin using copy number data. This is especially useful when clones are inferred from shallow single-cell DNA-seq, in which case the copy number state of each clone is known, but the precise SNV structure is unknown.

To assign cells to clones, `clonealign` makes the assumption that

$$ \text{gene expression} \propto \text{number of gene copies} $$

This is demonstrated in the figure below.

<img src="clonealign_figure.png" width="600" align="middle" />

Mathematically we have an $N \times G$ matrix $Y$ of raw gene expression counts (from RNA-seq) for $N$ cells and $G$ genes, where $y_{ng}$ is the counts to gene $g$ in cell $c$. We also have a $G \times C$ matrix $\Lambda$ of copy number variation for $C$ clones, where $\lambda_{gc}$ is the copy number of gene $g$ in clone $c$. We introduce a clone-assigning categorical variable $\pi_n$ for each cell, where

$$z_n = c \text{ if cell $n$ on clone $c$} $$

then `clonealign` models the conditional expected counts in a gene and cell as

$$ E[y_{ng} | z_n=c] = \frac{\lambda_{g,c} f(\mu_g) e^{\psi_n \cdot w_g}}{ \sum_{g'}\lambda_{g',c} f(\mu_{g'}) e^{\psi_n \cdot w_{g'}}} s_n  $$

where $s_n$ is a cell-specific size factor, $\mu_g$ is the per-chromosome expression (normalized so that $\mu_1 = 1$ for model identifiability), $f$ is a function that maps copy number to a multiplicative factor of expression, and $\psi$ and $w$ are cell and gene specific random effects. The noise distribution is assumed to be negative-binomial. Inference is performed using reparametrization-gradient variational inference to calculate a posterior distribution of the clone assignments $z_n$ and of all other model parameters.

# Installation

`clonealign` is built upon [Google's Tensorflow](https://www.tensorflow.org/) using the [Tensorflow R package provided by Rstudio](https://tensorflow.rstudio.com/). To install `tensorflow`, run 

```{r, eval=FALSE}
install.packages("tensorflow")
library(tensorflow)
install_tensorflow()
```

You can confirm the installation succeeded by running

```{r,eval=FALSE}
sess = tf$Session()
hello <- tf$constant('Hello, TensorFlow!')
sess$run(hello)
```

For more details see the [Rstudio page on tensorflow installation](https://tensorflow.rstudio.com/tensorflow/articles/installation.html).

`clonealign` can then be installed using the `devtools` package via

```{r, eval=FALSE}
devtools::install_github("kieranrcampbell/clonealign")
```

 
# Basic usage

## Data preparation

By default, `clonealign` requires two inputs:

* Gene expression data as raw counts. This can be in the form of a `SingleCellExperiment`, `SummarizedExperiment` or **cell by gene** `matrix`
* Copy number profiles for each clone and gene (where the genes **must** be the same as those measured in the expression data). This can be in the form of a `data.frame`, `DataFrame` or `matrix`

Bundled with the package is an example `SingleCellExperiment` for 100 genes and 200 cells:

```{r}
library(clonealign)
data(example_sce)
example_sce
```

This has raw integer counts in the `assays` slot as required for input to `clonealign`:

```{r}
assay(example_sce, "counts")[1:5, 1:5]
```


The CNV data is stored in the `rowData` of the `SingleCellExperiment` for 3 clones (A, B, and C) and crucially the same genes as the expression data:

```{r}
cnv_data <- rowData(example_sce)[, c("A", "B", "C")]
stopifnot(nrow(cnv_data) == nrow(example_sce)) # Make sure genes match up
head(cnv_data)
```



## Model fitting

The model is fitted with a basic call to `clonealign`, which prints the ELBO for each iteration (this can be turned off by setting `verbose = FALSE`):

```{r}
cal <- clonealign(example_sce, cnv_data)
```

```{r}
print(cal)
```

We can plot the ELBO to ensure convergence:

```{r}
qplot(seq_along(cal$elbo), cal$elbo, geom = c("point", "line")) +
  labs(x = "Iteration", y = "ELBO")
```

The maximum likelihood estimates of the clone assignments can be access through the `clone` slot:

```{r}
clones <- cal$clone
table(clones)
```

This can easily be added to the `SingleCellExperiment` for visualization with `scater`:

```{r}
example_sce$clone <- clones
example_sce <- normalize(example_sce)
plotPCA(example_sce, ncomponents = 3, colour_by = "clone")
```

The clone assignments in `clones` can then be used for the desired downstream analysis, such as differential expression or SNV analysis.

## Plotting results

The `plot_clonealign` function can be used to check the sanity of the fitted clones by ensuring that gene expression does correlate to the inferred copy number. For this we require the input `SingleCellExperiment`, copy number matrix, and clone assignments. Note that the `SingleCellExperiment` requires columns in `rowData` corresponding to the chromosome, start and end position of each feature (gene). These can conveniently be gathered using the `getBMFeatureAnnos` function in `scater`, e.g.

```{r, eval = FALSE}
sce <- getBMFeatureAnnos(sce, filters = "ensembl_gene_id",
                         attributes = c("ensembl_gene_id", "start_position", "end_position"),
                         feature_symbol = "hgnc_symbol",
                         feature_id = "ensembl_gene_id",
                         dataset = "hsapiens_gene_ensembl")
```

For now we'll set these to made-up values:

```{r}
gene_position <- as_data_frame(cnv_data) %>% 
  mutate(gene = seq_len(nrow(cnv_data))) %>% 
  arrange(A, B, C) %>% 
  mutate(position = seq_len(nrow(cnv_data))) %>% 
  arrange(gene) %>% 
  .$position

rowData(example_sce)$chromosome <- "1"
rowData(example_sce)$start_pos <- gene_position
rowData(example_sce)$end_pos <- gene_position
```

We can then plot the expression estimates using the `plot_clonealign` function:

```{r}
plot_clonealign(example_sce, cal$clone, cnv_data,
                chromosome = "1",
                chr_str = "chromosome",
                start_str = "start_pos",
                end_str = "end_pos")
```

where the `*_str` identifies the columns of `rowData(example_sce)` to look for the chromosome names and feature start and end positions.

## Evaluating the clone assignment

`clonealign` will assign single-cell RNA-seq to clones no matter how good the fit and no matter whether the clones actually exist in the expression data. As a consequence, it is important to evaluate the quality of fit which in practice we do by fitting the clones on a subset of genes and looking at how well the predicted expression profiles fits the observed expression data of the "held-out" genes. 

We perform his out-of-sample evaluation using the `evaluate_clonealign` function, which takes in the clonealign fit on the full dataset (along with the dataset) and both prints various metrics and returns a list containing various computed objects:

```{r, eval = TRUE}
ec <- evaluate_clonealign(example_sce, cnv_data, cal)
```

The function call prints several important metrics:

1. The mean square error (MSE) on the full dataset given the clone assignments, compared to the average MSE
under a null distribution found by permuting the clone assignments. If the observed MSE is greater than the null
MSE (either on average or a large percent of the time) then the fit is very poor and should be discarded
2. The agreement between the clone assignments on the full geneset and under the null geneset. If these are
very different then the algorithm is sensitive to the input genes and caution is advised
3. The MSE of held-out (out-of-sample) genes using a clonealign fit on an orthogonal test set of genes. Again, if
the observed MSE is greater than the null MSE (or approximately the same) then the fit is unreliable and should
be discarded.

These metrics are stored in a list returned by the function call. For details, see the help page at `?evaluate_clonealign`.

# Advanced options


## Controlling Variational Inference

Inference is performed using reparametrization gradient variational inference
which uses the evidence lower bound (ELBO) to monitor convergence. This is controlled using the
`rel_tol` parameter. When the difference 

$$ \Delta ELBO = \frac{ELBO_{\text{new}} - ELBO_{\text{old}}}{|ELBO_{\text{old}}|} $$
falls below `rel_tol`, the optimization algorithm is considered converged. The maximum number of iterations to acheive this is set using the `max_iter` parameter.

Inference is performed using Adam  (@kingma2014adam) which is controlled by three parameters:

* `learning_rate` the learning rate
* `rel_tol` the relative difference in the ELBO below which the optimization will be considered converged (see above)
* `max_iter` the maximum number of Adam iterations to perform (see above)


## Accessing posterior / maximum-likelihood parameter estimates

The object returned by a call to `clonealign` contains a `clone` slot for the maximum likelihood (ML) clone assignment for each cell. The ML estimates of the other parameters can be found in the `cal$ml_params` slot:

```{r}
names(cal$ml_params)
```

The slot `clone_probs` gives the probability that each cell is assigned to each clone:

```{r}
head(cal$ml_params$clone_probs)
```

while `mu` and `phi` give the maximum likelihood estimates of the $\mu$ and $\phi$ parameters from the model.


# Technical

```{r}
sessionInfo()
```

# References
